### PR TITLE
Implement realtime multiplayer

### PR DESCRIPTION
**Pull request by smoogipoo**
Implementing this as a one-pass thing since it's pretty self-contained.

### CODE REVIEW

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimePlayer.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+using osu.Framework.Allocation;
+using osu.Game.Online.Multiplayer;
+using osu.Game.Online.RealtimeMultiplayer;
+using osu.Game.Scoring;
+using osu.Game.Screens.Multi.Play;
+using osu.Game.Screens.Ranking;
+
+namespace osu.Game.Screens.Multi.RealtimeMultiplayer
+{
+    public class RealtimePlayer : TimeshiftPlayer
```
Comment (bdach) [inline osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimePlayer.cs:17]: This inheritance is maybe a bit ehh, but abstracting out further may be too much of muchness... Probably fine to leave be for now until it's a problem.

File: osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimePlayer.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            if (Token == null)
+                return; // Todo: Somehow handle token retrieval failure.
+
+            client.MatchStarted += onMatchStarted;
+            client.ResultsReady += onResultsReady;
+            client.ChangeState(MultiplayerUserState.Loaded);
+
+            while (!started)
+                Thread.Sleep(100);
```
Comment (bdach) [inline osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimePlayer.cs:46]: slight teeth pain at this but I suppose also maybe not super important for now - can this cause problems? maybe TODO it at least?

File: osu.Game/Online/Multiplayer/PlaylistItem.cs [hunk 3]
```diff
@@ -64,8 +64,8 @@ public PlaylistItem()
 
         public void MapObjects(BeatmapManager beatmaps, RulesetStore rulesets)
         {
-            Beatmap.Value = apiBeatmap.ToBeatmap(rulesets);
-            Ruleset.Value = rulesets.GetRuleset(RulesetID);
+            Beatmap.Value ??= apiBeatmap.ToBeatmap(rulesets);
+            Ruleset.Value ??= rulesets.GetRuleset(RulesetID);
```
Comment (bdach) [inline osu.Game/Online/Multiplayer/PlaylistItem.cs:68]: So `MapObjects()` won't ever overwrite a non-null beatmap anymore? The commit that introduced these (536df07) has a `(for testing)` annotation in the message - is this meant to be final? This affects the abstract `RoomManager`, and both multiplayer modes as a consequence.

File: osu.Game/Online/RealtimeMultiplayer/RealtimeMultiplayerClient.cs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        {
+            if (connection != null)
+                return;
+
+            connection = new HubConnectionBuilder()
+                         .WithUrl(endpoint, options =>
+                         {
+                             options.Headers.Add("Authorization", $"Bearer {api.AccessToken}");
+                         })
+                         .AddNewtonsoftJsonProtocol(options => { options.PayloadSerializerSettings.ReferenceLoopHandling = ReferenceLoopHandling.Ignore; })
```
Comment (bdach) [inline osu.Game/Online/RealtimeMultiplayer/RealtimeMultiplayerClient.cs:66]: Out of curiosity more than anything, does this serialiser setting actually matter? There are actual reference loops?

**Review by bdach — COMMENTED** [review]
Pretty hard (=large) diff to go through. Skimmed the new components without paying much attention - focused more on actual logic and changes to existing components.

File: osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimePlayer.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimePlayer.cs:17]: worth a TODO comment at least IMO.

File: osu.Game/Online/RealtimeMultiplayer/RealtimeMultiplayerClient.cs (hunk 4 above)
Comment (peppy) [reply osu.Game/Online/RealtimeMultiplayer/RealtimeMultiplayerClient.cs:66]: there was for spectator at least (beatmap -> beatmapset -> beatmaps). i think it's just a safe default at this point, either way.

File: osu.Game/Online/Multiplayer/PlaylistItem.cs (hunk 3 above)
Comment (smoogipoo) [reply osu.Game/Online/Multiplayer/PlaylistItem.cs:68]: Yes this is meant to be final. It's here for tests where the beatmap is set directly on `PlaylistItem`s.

File: osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimePlayer.cs (hunk 2 above)
Comment (smoogipoo) [reply osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimePlayer.cs]: You've brought up an interesting point though, which is that this should've had a timeout. I've changed it to using an MRE.

**Comment by peppy** [comment]
Trying to join an ended room gets the lobby into a permanently broken state:

File: osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimeRoomManager.cs [hunk 5]
```diff
@@ -38,10 +38,10 @@ protected override void LoadComplete()
         }
 
         public override void CreateRoom(Room room, Action<Room> onSuccess = null, Action<string> onError = null)
-            => base.CreateRoom(room, r => joinMultiplayerRoom(r, onSuccess), onError);
+            => base.CreateRoom(room, r => joinMultiplayerRoom(r, onSuccess, onError), onError);
```
Comment (peppy) [inline osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimeRoomManager.cs:41]: this is really hard to read :/

**Review by peppy — DISMISSED** [review]

**Comment by peppy** [comment]
Mostly happy with the state of changes (with a focus on changes to existing classes). For the new stuff, there's still many issues, a lot of which we are already tracking on a todo list. I think merging this ASAP and following up with PRs is the correct way forward from here on.

**Pull request merged**